pkg/abi/abidecode.go

```diff
@@ -94,7 +94,7 @@ func decodeABIElement(ctx context.Context, breadcrumbs string, block []byte, hea
 		}
 		headStart += headOffset
 		headPosition = headStart
-		return walkDynamicChildArrayABIBytes(ctx, "tup", breadcrumbs, block, headOffset, headPosition, component, component.tupleChildren)
+		return walkDynamicChildArrayABIBytes(ctx, "tup", breadcrumbs, block, headStart, headPosition, component, component.tupleChildren)
```

Comment:
This is very sensitive logic, so I want to make sure we fully understand the scope of the scenario that's broken.
I'll need to really do some deep thinking to understand that scenario, what we were doing before, and why this solves it.

Would you mind elaborating in the comment on the PR?

What I can see before:
- Before: `walkDynamicChildArrayABIBytes` was being called with:
   - `headStart = headOffset`
   - `headPosition = headStart+headOffset`
- After your PR: `walkDynamicChildArrayABIBytes` will be called with:
   - `headStart = headStart+headOffset`
   - `headPosition = headStart+headOffset` 

So before we were calculating two different values, and now we're passing the same value twice.

Clearly the case before worked in some scenarios, or the code would not have passed earlier testing.
However, clearly from this bug report there's another scenario where the previous calculation was wrong.

So to merge this one we (you and/or I) need to articulate exactly why the code before worked, the scenarios in which is was broken, and why the new calculation is right.

Sorry for the long text. I'm happy to spend the time to get to the point I can answer the above questions myself, just elaborating the task ahead of me in case you have already done it and just need to put down the detail.

Reply:
Another potential thing to check based on a quick look here, is that the test case you've got that we don't have a special case where `headStart == 0` and/or `headOffset == 0` on entry to this function. Those special cases are dangerous to base any assumptions around. Maybe adding a couple of extra test cases would also give extra confidence in the analysis:
- A tuple is an entry in a fixed array
- A tuple is an entry in a dynamic array
- A tuple is nested in another tuple

Reply:
Agreed, a lot goes through here and we need to be careful. Let me add those extra test cases.

The reason this solves the problem is just in the case where we need to a relative data offset.

Let's take the example from yesterday and map it to the function calls (this is how I got the debugger flow and understood what the problem was)


Full example
```
Blocks                                                           Index.  Value
0000000000000000000000000000000000000000000000000000000000000020 0       32 // offset for the list of tuples  
0000000000000000000000000000000000000000000000000000000000000001 32      1  // Number of tuples in the list (length)
0000000000000000000000000000000000000000000000000000000000000020 64      32 // Start position for the respective tuples
000000000000000000000000000000000000000000000000000000000000007b 96      123 // ID I gave the coupon 
0000000000000000000000000000000000000000000000000000000000000080 128     128 // Offset to length of string
000000000000000000000000000000000000000000000000000000000031079d 160     3213213 // _start of coupon I gave
0000000000000000000000000000000000000000000000000000000000005e4d 192     24141 // _end of coupon I gave
000000000000000000000000000000000000000000000000000000000000000b 224     11 // length of string in bytes
7468697369736d7975726c000000000000000000000000000000000000000000 256    thisismyurl // String starts here 128 + 128 
```

Walkthrough of example:

Call to `DecodeABIDataCtx` with **offset 0** 
1) First we create a TypeComponentTree which in this case is a TupleComponent with 4 parameters defined below

`
github.com/hyperledger/firefly-signer/pkg/abi.ParameterArray len: 4, cap: 4, [*{Name: "_tokenId", Type: "uint256", InternalType: "", Components: github.com/hyperledger/firefly-signer/pkg/abi.ParameterArray len: 0, cap: 0, nil, Indexed: false, parsed: *github.com/hyperledger/firefly-signer/pkg/abi.typeComponent nil},*{Name: "_tokenURL", Type: "string", InternalType: "", Components: github.com/hyperledger/firefly-signer/pkg/abi.ParameterArray len: 0, cap: 0, nil, Indexed: false, parsed: *github.com/hyperledger/firefly-signer/pkg/abi.typeComponent nil},*{Name: "_startDate", Type: "uint256", InternalType: "", Components: github.com/hyperledger/firefly-signer/pkg/abi.ParameterArray len: 0, cap: 0, nil, Indexed: false, parsed: *github.com/hyperledger/firefly-signer/pkg/abi.typeComponent nil},*{Name: "_endDate", Type: "uint256", InternalType: "", Components: github.com/hyperledger/firefly-signer/pkg/abi.ParameterArray len: 0, cap: 0, nil, Indexed: false, parsed: *github.com/hyperledger/firefly-signer/pkg/abi.typeComponent nil}]
`

2) We walk through the ABI bytes at **offset=0** so **headStart = 0** and **headPosition=0**
   - In `fn walkDynamicChildArrayABIBytes` for each children in the top DynamicArray we call decodeABIElement
   - In `fn decodeABIElement ` hit the big switch statement in and we go inside the `DynamicArrayComponent` case
   - Within the case we decode the first 32 bytes to get the headOffset (i.e where the list starts) in this case **32**
   - and we call `fn walkDynamicChildArrayABIBytes` with `headStart + headOffset` == 32 
   - 

3) We find the length of the dynamic array in `fn decodeABIDynamicArrayBytes` at **dataOffset = 32**
  - We decode the next 32 bytes to find the length of the list (in our case the number of tuples) and we increase the dataOffset by 32
  - For each children in that array we decode it, so we call `fn decodeABIElement` and we hit the `TupleComponent` case
 
 
 4) Decode the Tuple Component - **headStart = 64** and **headPosition = 64**
  - We first get the offset to the start of the tuple by calling `decodeABILength` for the next 32 bytes and get 32
  - With that value we increase the headStart so **headStart += headOffset** == 96 (which is the correct place to where the start of the tuple is 
  - We then call `walkDynamicChildArrayABIBytes` at that offset because we want to walk the tuple component now.
  **Note**: Not sure why. we passed the headOffset before , maybe we used it in a different way?
  
 5) Let's walk the children of the tuple component - **headStart = 96** and **headPosition = 96**
  - Now for each children:
    - In the case of the `uint256` simple, just decode the next 32 bytes and easy ->  **headStart = 96** and **headPosition = 128** 
    **Important note** - inside the for loop we increment the `headPosition += childHeadBytes` but the headStart needs to stay the same!
    - In the case of the `string` it's trickier  **headStart = 96** and **headPosition = 128** , we end up in `fn decodeABIBytes`
      - As it's a variable length type, we get the  first 32 bytes by calling `fn decodeABILength` which gives you the offset to the string -> in this case 128 
      - This is where we know want to get the bytes for this string 
         - Before we were doing `headStart + dataOffset` since headStart was 32 this was 160 which is not where the string is and thus the error
         - In this PR I'm doing the same but headStart is 96 which takes us to 224 where the bytes are but **it's wrong!** we don't want this! I only got lucky because the string is in the second place in the tuple 
         - **_What we need to do is add up the current position `headPosition` to the `dataOffset` which is the position of the string and then remove 32 to get the location for the length of the string_**
    - Don't need to explain the rest as are uint256 which are simple ones.
   
   That was long so happy to have a zoom over it :)